feat: SSR `split` mode (#7220)
withastro/astro: .changeset/wet-readers-join.md

---
'astro': minor
---

Shipped a new SSR mode, called `split`. 

> When enabled, Astro will emit a file for each page, which will render one single page.

Reviewer: ```suggestion
When enabled, Astro will "split" the single `entry.mjs` file and instead emit a separate file to render each individual page during the build process.
```
I'm assuming this is where the name "split" came from?  :smile: It would be good to have something to connect this name to what it does.

> These files will be emitted inside `dist/pages`, and they will look like this:

Reviewer: ```suggestion
These files will be emitted inside `dist/pages`, mirroring your page files in `src/pages/`, for example:
```
This is the wording we went with in docs, and is more descriptive than "they will look like this" so I prefer something like this!

Author: Mirroring is not technically correct because the name of the files will look different. The only thing that technically we mirror is the **directory structure**. The name of the file will be different, e.g.: `blog.astro` -> `entry.blog.astro.mjs`

Reviewer: OK, what about something like "mirroring the directory structure of your page files in `src/pages/`"... ?

```
├── pages
│   ├── blog
│   │   ├── entry._slug_.astro.mjs
│   │   └── entry.about.astro.mjs
│   └── entry.index.astro.mjs
```

> To enable and customise this mode, new options are now available:

Reviewer: ```suggestion
To enable, set `build.split: true` in your Astro config:
```
I don't think this is customizable? Only enable-able? I would mention that this is a build option, one way or another here.

> ```js

Reviewer: ```suggestion
```js
// src/astro.config.mjs
```
Let's add a file title, whatever the convention is here!

---
'astro': minor
---

Shipped a new SSR build configuration mode: `split`. 
When enabled, Astro will "split" the single `entry.mjs` file and instead emit a separate file to render each individual page during the build process.

> These files will be emitted inside `dist/pages`, and they will look like this:

Reviewer: ```suggestion
These files will be emitted inside `dist/pages`, mirroring the directory structure of your page files in `src/pages/`, for example:
```
Updated my suggestion to be more precise about the mirroring. Would this work?

---
'astro': minor
---

Shipped a new SSR mode, called `serverless`. 
When enabled, Astro will emit a file for each page, which will render one single page.

> These files will be emitted inside `dist/pages`, and they will look like this:

Reviewer 2: The changeset docs here also needs to be updated.

---
'astro': minor
---

Shipped a new SSR mode, called `serverless`. 
When enabled, Astro will emit a file for each page, which will render one single page.

These files will be emitted inside `dist/pages`, and they will look like this:

```
├── pages
│   ├── blog
│   │   ├── entry._slug_.astro.mjs
│   │   └── entry.about.astro.mjs
│   └── entry.index.astro.mjs
```

To enable and customise this mode, new options are now available:

```js
export default defineConfig({
    output: "server",
    adapter: node({
        mode: "standalone"
    }),
    build: {
        ssrMode: "serverless",
        serverlessEntryPrefix: "main"
    }
})
```

- `ssrMode` accepts two values, `"server"` or `"serverless"`. Default value, `"server"`. 

> - `serverlessEntryPrefix` allows to change the prefix of a serverless

Reviewer 2: I wonder if `serverless` is the right term here. The previous build output can and are also used on serverless platforms. SvelteKit's Vercel adapter calls it ["split"](https://kit.svelte.dev/docs/adapter-vercel#deployment-configuration), since we're doing on the build-level, maybe we can call it `build.pageCodeSplit`? Vite has a similar option called `vite.build.cssCodeSplit`.

I don't really understand the `serverlessEntryPrefix` part yet, but maybe I will when I finish reviewing this PR!

Reviewer 3: are you proposing `build.pageCodeSplit` be a boolean or what?

Reviewer 2: Yeah, it would be a boolean which defaults to false.

Author: I probably misunderstand the meaning, because in my head it seems we split the code of a page in multiple files. 

But it makes sense to rename the option to something that's more familiar to the majority of users
